### PR TITLE
fix(auth): replace bespoke wildcard matching with glob-based redirect…

### LINKED ISSUE #1220: Support Supabase-style glob patterns in allowedRedirectUrls (path/recursive wildcards)
## Background

The `allowedRedirectUrls` validation in `backend/src/services/auth/auth-config.service.ts:295` (`validateRedirectUrl`) only supports a narrow form of wildcarding: an entry is treated as a wildcard pattern **only if it contains the literal substring `*.`**, and even then the wildcard only applies to the hostname's leading subdomain.

```ts
// auth-config.service.ts:295-323 (validateRedirectUrl)
return allowedRedirectUrls.some((item) => {
  if (!item.includes('*.')) {
    return this.normalizeUrl(item) === targetUrl;   // exact-match fallback
  }
  // ... only host-level *. handled
});
```

Anything that doesn't contain `*.` falls into the exact-string-match branch, so non-host patterns silently never match anything real. This is undocumented and the failure mode looks like a bug rather than a misconfigured pattern.

## What works today

- `https://*.example.com` → matches `https://app.example.com`, `https://staging.example.com`

## What doesn't work today (silently)

- `*` alone — only matches the literal string `*`
- `https://example.com/*` — **path wildcards are unsupported**
- `https://example.com/**` — same
- `https://*foo.example.com` — partial subdomain wildcards
- `https://example.*` — TLD wildcards
- `*.example.com` matching the apex `example.com` itself (code requires `targetUrl.hostname.endsWith('.' + baseDomain)`, so the apex doesn't match)
- Even with `*.`: protocol mismatch (http vs https), port mismatch, or any non-`/` path that's not exactly equal will fail

## Customer impact

A customer (mentormates.ai) wanted `mentormates.ai/*` to allow any redirect path on their domain. This is impossible to express in the current validator — they'd have to enumerate every callback path explicitly.

Coming from Supabase this is a regression, since Supabase supports glob patterns directly:

| Pattern | Supabase | InsForge today |
|---|---|---|
| `*.example.com` (subdomain) | ✅ | ✅ |
| `example.com/*` (path, single segment) | ✅ | ❌ silent fail |
| `example.com/**` (path, recursive) | ✅ | ❌ silent fail |
| `*.example.com/auth/*` | ✅ | ❌ partial only |
| `example.com/?session=*` | ✅ | ❌ |

Supabase reference: https://supabase.com/docs/guides/auth/redirect-urls

## Proposal

Replace the bespoke `*.` substring check with a real glob matcher (e.g., `picomatch` or `minimatch`), using `.` and `/` as path separators — Supabase-compatible defaults so customers migrating from Supabase get the same behavior.

```ts
// pseudocode
import picomatch from 'picomatch';

async validateRedirectUrl(urlStr: string): Promise<boolean> {
  const config = await this.getAuthConfig();
  const allowed = config.allowedRedirectUrls;
  if (!allowed?.length) return true;
  
  const target = this.normalizeUrl(urlStr);
  if (!target) return false;
  
  return allowed.some(pattern => {
    const normalizedPattern = this.normalizeUrl(pattern) || pattern;
    // Glob match across the whole URL with . and / as separators
    return picomatch(normalizedPattern, { 
      dot: true, 
      // separators not built-in to picomatch — may need to write a small shim
    })(target);
  });
}
```

Pattern semantics (matching Supabase):
- `*` — matches any sequence of non-separator chars (separators = `.` and `/`)
- `**` — matches any chars including separators
- `?` — matches a single non-separator char
- `[!a-z]` — character ranges

## Acceptance criteria

- [ ] `validateRedirectUrl` accepts Supabase-style glob patterns
- [ ] Existing `*.host.com` patterns still work (back-compat for current customers)
- [ ] `example.com/*` matches `example.com/foo` but NOT `example.com/foo/bar`
- [ ] `example.com/**` matches both
- [ ] `*.example.com` matches both `app.example.com` AND optionally `example.com` itself (decide: match apex when explicitly requested via separate config flag, or only match subdomains for safety)
- [ ] Protocol/port match remains strict (don't loosen these — security boundary)
- [ ] Unit tests covering all the cases in the table above
- [ ] Docs in `docs/auth/redirect-urls.mdx` describing supported patterns + examples

## File pointers

- `backend/src/services/auth/auth-config.service.ts:273-326` — `validateRedirectUrl`
- `packages/shared-schemas/src/auth.schema.ts` — `allowedRedirectUrls` field
- Existing dashboard UI in `packages/dashboard/src/features/auth/` — settings page, may want to add inline validation/preview

## Security notes

- Glob patterns are powerful — make sure `*` cannot escape the host. e.g., `*.example.com` shouldn't allow `evil.com.example.com` style domain confusion. Picomatch with proper anchoring handles this correctly.
- Apex matching for `*.example.com` should be explicit — silently allowing the apex when only subdomains are configured is a footgun.
- Document the `**` operator clearly — customers should understand that `https://example.com/**` allows ANY path on that domain, including admin/internal routes.

---
_Filed based on customer feedback (mentormates.ai). See Insforge Tech use chat 2026-05-07 for original report._

### PR DESCRIPTION
**Pull request by Davidson3556**
## Summary

Closes #1220. Replaces the bespoke `*.` substring check in `validateRedirectUrl`
with a real glob matcher (`picomatch`) so `allowedRedirectUrls` accepts
Supabase-style patterns instead of silently failing on anything that's not
`https://*.host.com`.

- `validateRedirectUrl` (in [auth-config.service.ts](backend/src/services/auth/auth-config.service.ts)) now matches via `picomatch` after normalising both pattern and target through the `URL` class. Glob meta-chars (`*`, `**`, `?`, `[…]`) are placeholder-swapped before normalisation so `new URL()` doesn't reject or percent-encode them.
- `allowedRedirectUrlsRegex` (in [auth.schema.ts](packages/shared-schemas/src/auth.schema.ts)) widened to accept `*` / `**` / `?` / `[…]` in the host and path; protocol stays strict (no globbing the scheme; `javascript:` / `data:` / etc. still rejected via the existing negative lookahead).
- Back-compat preserved: existing `*.host.com` patterns continue to match arbitrarily deep subdomains (the legacy matcher used `endsWith('.' + base)` which did the same).
- Apex match for `*.example.com` is intentionally **not** silently included — you must list `example.com` separately. Protocol and port remain strict — wildcards never cross those boundaries.
- New unit tests in [auth-config.service.test.ts](backend/tests/unit/auth-config.service.test.ts) cover every row of the issue's pattern table (subdomain, single-segment path, recursive path, combined, query-string, protocol/port mismatch, domain-confusion, malformed input, custom schemes).
- Docs added at [docs/core-concepts/authentication/redirect-urls.mdx](docs/core-concepts/authentication/redirect-urls.mdx) describing supported patterns, security caveats, and Supabase migration notes; wired into [docs/docs.json](docs/docs.json).

## How did you test this change?

- [x] `cd backend && npx vitest run tests/unit/auth-config.service.test.ts` — 32/32 passing
- [x] `npx tsc --noEmit` in `backend` and `packages/shared-schemas` — clean
- [x] `npx prettier --check` on all touched files — clean
- [ ] Manual end-to-end via dashboard: configure `https://*.example.com/auth/*` and `https://example.com/?session=*` in Auth → Settings, confirm a real OAuth/email-verify redirect against each
- [ ] Verify existing customers with `https://*.host.com` patterns still pass (regression check)

---
## Summary by cubic
Replaced bespoke wildcard checks with glob-based redirect URL validation using `picomatch`, enabling Supabase-style patterns across host, path, and query. Keeps protocol/port strict, preserves deep subdomain behavior, and adds comprehensive tests.

- Bug Fixes
  - Match via `picomatch` over normalized URLs; supports `*`, `**`, `?`, `[…]`; `**` also matches the root URL.
  - Preserve `*.example.com` deep-subdomain behavior; apex must be added separately.
  - Enforce strict protocol and port; supports custom schemes; globs never cross scheme/port.
  - Handle IPv6 hosts correctly by treating `[::1]`-style brackets as literals to prevent over-matching.
  - Expand allowed-pattern regex to allow glob chars outside protocol, require a non-empty host, and block dangerous schemes; unparseable patterns are rejected at runtime.
  - Make normalizeUrl and normalizePattern static for safer reuse.

- Dependencies
  - Add `picomatch` and `@types/picomatch`.

<sup>Written for commit 14e01cf28dccc48a46bd762bf7fa06564d560b67. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Redirect URL validation now supports advanced glob patterns (*, **, ?, character classes) with improved normalization and matching behavior.

* **Documentation**
  * Added a new guide page for configuring redirect URLs under Authentication.

* **Tests**
  * Expanded unit tests to cover glob semantics, edge cases (IPv6, custom schemes, malformed inputs), and schema validation for allowed redirect patterns.

[Review Change Stack](https://app.coderabbit.ai/change-stack/InsForge/InsForge/pull/1235)

### CODE REVIEW

**Comment by Davidson3556** [comment]
@tonychang04 kindly review

**Review by jwfing — DISMISSED** [review]
LGTM, approved.

**Review by jwfing — COMMENTED** [review]
## Summary

Solid, well-scoped bug fix that replaces a fragile substring-based wildcard with a proper glob engine (`picomatch`). The implementation is correct, backward-compatible, and thoroughly tested. No blocking issues found.

## Requirements context

No matching spec or plan was found under `/docs/superpowers/specs/` or `/docs/superpowers/plans/` for redirect-URL validation. Assessment is based on PR description, issue #1220, and the existing `validateRedirectUrl` contract.

---

## Findings

### Critical

_(none)_

---

### Suggestion

**[Functionality] `normalizePattern` fallback has a latent case-sensitivity mismatch — `auth-config.service.ts:305-311`**

The fallback branch (reached when `normalizeUrl(safe)` returns `null`) lowercases the entire pattern string before restoring glob tokens:

```ts
let fallback = pattern.toLowerCase().replace(/\/$/, '');
```

But the `normalizedTarget` passed in from `matchesGlobPattern` is produced by `new URL(urlStr).href`, which lowercases only the *hostname* — the path preserves its original case (per the URL spec). If a path component differs in case between pattern and target, the subsequent `picomatch.isMatch` call can fail to match even though it should.

**Blast radius is low** — this fallback is dead code for any pattern that passes `allowedRedirectUrlsRegex` (all valid inputs), because schema validation guarantees the pattern is a parseable URL. But it is silent incorrect behavior for any edge cases that somehow reach it (e.g. a code path that sets `allowedRedirectUrls` without going through the schema). Options:

- Replace the lowercased fallback with a simple `return pattern.replace(/\/$/, '')` (keep original case, since `picomatch.isMatch` with `dot: true` respects case by default, matching the URL spec behavior).
- Or remove the fallback entirely and return an unmatchable sentinel string (e.g. `'\0'`) — this makes the rejection explicit and removes a dead branch.

**[Software Engineering] Fallback path has no direct test coverage — `auth-config.service.ts:305-311`**

The fallback in `normalizePattern` is an instance-private method, so it's only tested indirectly through `validateRedirectUrl`. No test currently exercises the fallback branch (a pattern that remains unparseable after placeholder substitution). Consider:
- Adding an integration-style test where the allowlist entry is a deliberately malformed pattern (bypassing schema validation via the mock), to pin the fallback's behavior.
- Or, as above, eliminating the fallback so there is no branch to miss.

---

### Information

**[Software Engineering] `normalizeUrl` and `normalizePattern` could be `private static` — `auth-config.service.ts:263-321`**

Neither method uses instance state — `normalizeUrl` only calls `new URL()` and `normalizePattern` only calls `this.normalizeUrl`. Making both `private static` would make the absence of side-effects explicit and allows calling them without a service instance in future tests. Minor style note only.

**[Software Engineering] Test gap: `**` vs root path — `auth-config.service.test.ts` (recursive path wildcard block)**

The `**` tests cover `/foo` and `/foo/bar/baz`, but not `https://example.com/**` matched against `https://example.com/` (a trailing-slash root). After normalization, `normalizeUrl('https://example.com/')` strips the trailing slash to `https://example.com`, while the pattern becomes `https://example.com/**`. Whether picomatch treats `https://example.com` as matching `https://example.com/**` (i.e., is an empty segment after the literal `/` accepted by `**`?) is worth pinning with an explicit test.

**[Functionality / Docs] Permissive-default callout could use `<Warning>` — `docs/core-concepts/authentication/redirect-urls.mdx:15-19`**

The current `<Info>` block describes the empty-allowlist behavior:

> If no redirect URLs are configured (empty list), InsForge allows **all** redirect URLs. This is intentional…

This is a high-impact footgun in production — any `redirectTo` value, including crafted ones, will pass validation until the allowlist is populated. Surfacing it as `<Warning>` would better signal to operators that this is a configuration step they must not skip before going live.

---

## Verdict

**approved** _(informational — no Critical findings; human approval required via the approve flow)_

All four review dimensions checked:
- **Software engineering** — 32 tests, excellent coverage across all documented pattern types; test structure is clean and readable.
- **Functionality** — back-compat for `*.host.com` is preserved and verified; protocol and port strictness are correctly enforced by keeping them as literal characters in the normalized string before glob matching; apex-domain exclusion is intentional and tested.
- **Security** — `picomatch` is a widely-used, reputable library (part of the micromatch/chokidar ecosystem) with no known CVEs at this version; no secrets logged or returned; no auth checks weakened; placeholder restoration does not introduce regex-replacement injection since no glob token begins with `$`.
- **Performance** — `picomatch.isMatch` uses an internal compiled-pattern cache keyed by pattern string, so repeated calls with the same allowlist patterns incur no re-compilation overhead. `allowedRedirectUrls` lists are expected to be small; no N+1 concern.

File: docs/core-concepts/authentication/redirect-urls.mdx [hunk 1]
```diff
@@ -0,0 +1,138 @@
+---
```
Comment (tonychang04) [inline docs/core-concepts/authentication/redirect-urls.mdx]: can you not change our doc right now? i think this is good but docs maybe we can add it ourselves

**Comment by tonychang04** [comment]
@cubic-dev-ai review

File: docs/docs.json [hunk 2]
```diff
@@ -47,7 +47,8 @@
               {
                 "group": "Authentication",
                 "pages": [
-                  "core-concepts/authentication/architecture"
+                  "core-concepts/authentication/architecture",
+                  "core-concepts/authentication/redirect-urls"
```
Comment (tonychang04) [inline docs/docs.json]: can we remove the docs as mentioned befor 😂

Comment (Davidson3556) [reply docs/docs.json]: okay

**Comment by Davidson3556** [comment]
@tonychang04  kindly review

**Review by tonychang04 — APPROVED** [review]

**Comment by tonychang04** [comment]
🚀merge!
